src/platform/windows/display_uwp.cpp

```diff
@@ -0,0 +1,247 @@
+    auto interop_factory = winrt::get_activation_factory<winrt::GraphicsCaptureItem, IGraphicsCaptureItemInterop>();
+    return interop_factory->CreateForMonitor(mon, winrt::guid_of<winrt::Windows::Graphics::Capture::IGraphicsCaptureItem>(), winrt::put_abi(item));
+  }
+
+  uwp_capture_t::uwp_capture_t() {
+    InitializeCriticalSection(&frame_lock);
+    InitializeConditionVariable(&frame_present_cv);
+  }
+
+  uwp_capture_t::~uwp_capture_t() {
+    capture_session.Close();
+    frame_pool.Close();
+    item = nullptr;
+    capture_session = nullptr;
+    frame_pool = nullptr;
+  }
+
+  int
+  uwp_capture_t::init(display_base_t *display, const ::video::config_t &config) {
+    HRESULT status;
+    dxgi::dxgi_t dxgi;
+    winrt::com_ptr<::IInspectable> d3d_comhandle;
+    if (!winrt::GraphicsCaptureSession::IsSupported()) {
+      BOOST_LOG(error) << "Screen capture is not supported on this device for this release of Windows!"sv;
+      return -1;
+    }
+    if (FAILED(status = display->device->QueryInterface(IID_IDXGIDevice, (void **) &dxgi))) {
+      BOOST_LOG(error) << "Failed to query DXGI interface from device [0x"sv << util::hex(status).to_string_view() << ']';
+      return -1;
+    }
+    if (FAILED(status = winrt::CreateDirect3D11DeviceFromDXGIDevice(*&dxgi, d3d_comhandle.put()))) {
+      BOOST_LOG(error) << "Failed to query WinRT DirectX interface from device [0x"sv << util::hex(status).to_string_view() << ']';
+      return -1;
+    }
+
+    uwp_device = d3d_comhandle.as<winrt::IDirect3DDevice>();
+    DXGI_OUTPUT_DESC output_desc;
+    display->output->GetDesc(&output_desc);
+    if (FAILED(status = create_item_for_monitor(output_desc.Monitor, item))) {
+      BOOST_LOG(error) << "Failed to activate GraphicsCaptureItem for monitor [0x"sv << util::hex(status).to_string_view() << ']';
+      return -1;
+    }
+    if (config.dynamicRange && display->is_hdr())
+      display->capture_format = DXGI_FORMAT_R16G16B16A16_FLOAT;
+    else
+      display->capture_format = DXGI_FORMAT_B8G8R8A8_UNORM;
+
+    frame_pool = winrt::Direct3D11CaptureFramePool::CreateFreeThreaded(uwp_device, static_cast<winrt::Windows::Graphics::DirectX::DirectXPixelFormat>(display->capture_format), 2, item.Size());
+    capture_session = frame_pool.CreateCaptureSession(item);
+    frame_pool.FrameArrived({ this, &uwp_capture_t::on_frame_arrived });
+    capture_session.IsBorderRequired(false);
+    capture_session.IsCursorCaptureEnabled(true);
+    capture_session.StartCapture();
+    return 0;
+  }
+
+  void
+  uwp_capture_t::on_frame_arrived(winrt::Direct3D11CaptureFramePool const &sender, winrt::IInspectable const &) {
+    // this PRODUCER runs in a separate thread spawned by the frame pool. to
+    // retain parity with the original API, the frame will be consumed by
+    // the capture thread, not this one.
+
+    auto frame = sender.TryGetNextFrame();
+    if (frame != nullptr) {
+      EnterCriticalSection(&frame_lock);
+      capture_frame = frame;

```

Comment:
It looks like we will leak the previous frame if another frame arrives before the capture thread consumed the last one.

Maybe we could have 2 frame member variables:
- `producer_frame` written by `on_frame_arrived()` (closing the old one if it still exists)
- `consumer_frame` moved from `producer_frame` in `next_frame()` then closed in `release_frame()`. 

We can avoid synchronizing `consumer_frame` with `frame_lock` since it's guaranteed to only be touched on a single thread (other than when we move `consumer_frame` into it in `next_frame()`).

Reply:
<pedantry>Frames don't leak. They're managed by the frame pool, and merely borrowed by the application.</pedantry>

I'll make a change based on this suggestion, though, since I don't want a frame to be overwritten by the on_frame_arrived callback here while it's being processed in `snapshot`.